### PR TITLE
[json] Add Rubix\ML\Other\Helpers\JSON

### PR DESCRIPTION
**Pull request by simplechris**
##### SUMMARY:

- Add `Rubix\ML\Other\Helpers\JSON` and throw more detailed exception messages if malformed JSON is encountered. 

Small enhancement reconstructed from some local stashes that I had laying around. Let me know if you think it's worth merging 🤷‍♂️

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -34,6 +34,7 @@
     ],
     "require": {
         "php": ">=7.2",
+        "ext-json": "*",
```
Comment (andrewdalpino) [inline composer.json:37]: Is this line necessary if we require PHP 7.2+?

https://www.php.net/manual/en/json.installation.php#json.installation

File: src/Other/Helpers/JSON.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                return $value;
+            case JSON_ERROR_SYNTAX:
+                throw new RuntimeException('Malformed JSON Provided. Syntax Error.');
+            case JSON_ERROR_DEPTH:
+                throw new RuntimeException('Maximum stack depth exceeded.');
+            case JSON_ERROR_STATE_MISMATCH:
+                throw new RuntimeException('Underflow or the modes mismatch.');
+            case JSON_ERROR_CTRL_CHAR:
+                throw new RuntimeException('Unexpected control character found.');
+            case JSON_ERROR_UTF8:
```
Comment (andrewdalpino) [inline src/Other/Helpers/JSON.php:63]: Some of the possible errors are not listed here such as NAN and INF. Should we catch those errors or did you intend for the default case to handle those?

https://www.php.net/manual/en/function.json-last-error.php#refsect1-function.json-last-error-returnvalues

File: src/Other/Helpers/JSON.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     * @param int $depth
+     *
+     * @throws \RuntimeException
+     *
+     * @return string
+     */
+    public static function encode($value, int $options = self::DEFAULT_OPTIONS, int $depth = self::DEFAULT_DEPTH) : string
+    {
+        $data = json_encode($value, $options, $depth);
+        if ($data === false) {
```
Comment (andrewdalpino) [inline src/Other/Helpers/JSON.php:92]: Nice explicit check for false here rather than falsey

**Review by andrewdalpino — COMMENTED** [review]
I think this will be quite useful ... I had a couple of questions about some particular JSON errors that we aren't catching in the switch statement. Other than that looks good to me!

File: composer.json (hunk 1 above)
Comment (simplechris) [reply composer.json:37]: Yeah, I was surprised by this! I've never _ever_ used an installation of php without json enabled. 

I added this to satisfy my IDE that complains that `'ext-json' is not listed in composer.json` 

It seems that although ext-json is enabled by default you can still compile php without it with using the `--disable-json` flag, so to use json you should technically announce it as a project dependency. 🤷‍♂️  


---
**EDIT:** This flag is removed in PHP8 meaning json will always be available (https://externals.io/message/109783)

File: src/Other/Helpers/JSON.php (hunk 2 above)
Comment (simplechris) [reply src/Other/Helpers/JSON.php:63]: 👍 Yeah, let me add those in now

**Review by simplechris — COMMENTED** [review]
@andrewdalpino updated and included all the json error states: ([diff](https://github.com/RubixML/RubixML/compare/84736969692becafb1dbf52163a0d1c93f66355c..20573cf1f6d1e5af8f59e88e31b798060f810a4d))

If you want me to remove the `ext-json` thing then totally happy to, it seems more like a theoretical edge-case and I only became aware of it via IDE notifications.

File: composer.json (hunk 1 above)
Comment (andrewdalpino) [reply composer.json:37]: For shizzle, thanks for checking that out

**Review by andrewdalpino — APPROVED** [review]

**Comment by andrewdalpino** [comment]
Thanks again @simplechris, very helpful!

**Pull request merged**